presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,129 @@
+    {
+        return treeRewriter.defaultRewrite(node, context);
+    }
+
+    @Override
+    public Expression rewriteGroupingOperation(GroupingOperation node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        Expression rewrittenExpression = rewriteGroupingOperationToExpression(node, queryNode);
+        if (rewrittenExpression instanceof FunctionCall) {
+            FunctionCall rewrittenFunctionCall = (FunctionCall) rewrittenExpression;
+            List<TypeSignatureProvider> functionTypes = Arrays.asList(
+                    new TypeSignatureProvider(BIGINT.getTypeSignature()),
+                    new TypeSignatureProvider(ListLiteralType.LIST_LITERAL.getTypeSignature()),
+                    new TypeSignatureProvider(ListLiteralType.LIST_LITERAL.getTypeSignature())
+            );
+            resolveFunction(rewrittenFunctionCall, functionTypes, metadata.getFunctionRegistry());
+        }
+
+        return rewrittenExpression;
+    }
+
+    private Expression rewriteGroupingOperationToExpression(GroupingOperation expression, QuerySpecification node)
+    {
+        List<Expression> columnReferences = ImmutableList.copyOf(analysis.getColumnReferences());
+        List<Expression> groupingOrdinals = expression.getGroupingColumns().stream()
+                .map(columnReferences::indexOf)
+                .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
+                .collect(toImmutableList());
+
+        List<List<Expression>> groupingSetOrdinals;
+        ImmutableList.Builder<List<Expression>> groupingSetOrdinalsBuilder = ImmutableList.builder();
+        for (List<Expression> groupingSet : analysis.getGroupingSets(node)) {
+            groupingSetOrdinalsBuilder.add(groupingSet.stream()
+                    .map(columnReferences::indexOf)
+                    .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
+                    .collect(toImmutableList()));
+        }
+        groupingSetOrdinals = groupingSetOrdinalsBuilder.build();
+
+        checkState(node.getGroupBy().isPresent(), "GroupBy node must be present");
+        List<GroupingElement> groupingElements = node.getGroupBy().get().getGroupingElements().stream()
+                .filter(element -> element instanceof GroupingSets || element instanceof Rollup || element instanceof Cube)
+                .collect(toImmutableList());
+        Expression firstArgument;
+        // No GroupIdNode and a GROUPING() operation imply a single grouping, which
+        // means that any columns specified as arguments to GROUPING() will be included
+        // in the group and none of them will be aggregated over. Hence, re-write the
+        // GroupingOperation to a constant literal of 0.
+        // See SQL:2011:4.16.2 and SQL:2011:6.9.10.
+        if (groupingElements.isEmpty() && !groupIdSymbols.containsKey(node)) {
```

Comment:
Ya it can be empty in the case of `grouping()` with either `GROUPING SETS`, `ROLLUP` or `CUBE`. As I point out in https://github.com/prestodb/presto/pull/6704#discussion_r106288356, this code section is not well written so I'll refactor it.